pkg/storegateway/expanded_replication.go

```diff
@@ -0,0 +1,107 @@
+	}
+
+	return nil
+}
+
+// ReplicatedBlock is a TSDB block that may be eligible to be synced to more store-gateways
+// than the configured replication factor based on metadata about the block.
+type ReplicatedBlock interface {
+	GetMinTime() time.Time
+	GetMaxTime() time.Time
+}
+
+// QueryableReplicatedBlock is a TSDB block that may be eligible to be queried from more
+// store-gateways than the configured replication factor based on metadata about the block.
+type QueryableReplicatedBlock interface {
+	ReplicatedBlock
+
+	GetUploadedAt() time.Time
+}
+
+// ExpandedReplication determines if a TSDB block is eligible to be sync to and queried from more
+// store-gateways than the configured replication factor based on metadata about the block.
+type ExpandedReplication interface {
+	// EligibleForSync returns true if the block can be synced to more than the configured (via
+	// replication factor) number of store-gateways, false otherwise.
+	EligibleForSync(b ReplicatedBlock) bool
+
+	// EligibleForQuerying returns true if the block can be safely queried from more than the
+	// configured (via replication factor) number of store-gateways, false otherwise.
+	EligibleForQuerying(b QueryableReplicatedBlock) bool
+}
+
+func NewNopExpandedReplication() *NopExpandedReplication {
+	return &NopExpandedReplication{}
+}
+
+// NopExpandedReplication is an ExpandedReplication implementation that always returns false.
+type NopExpandedReplication struct{}
+
+func (n NopExpandedReplication) EligibleForSync(ReplicatedBlock) bool {
+	return false
+}
+
+func (n NopExpandedReplication) EligibleForQuerying(QueryableReplicatedBlock) bool {
+	return false
+}
+
+func NewMaxTimeExpandedReplication(maxTime time.Duration, gracePeriod time.Duration) *MaxTimeExpandedReplication {
+	return &MaxTimeExpandedReplication{
+		maxTime:     maxTime,
+		gracePeriod: gracePeriod,
+		now:         time.Now,
+	}
+}
+
+// MaxTimeExpandedReplication is an ExpandedReplication implementation that determines
+// if a block is eligible for expanded replication based on how recent its MaxTime (most
+// recent sample) is. An upload grace period can optionally be used during queries to ensure
+// that blocks are not expected to be replicated to store-gateways that may not have yet had
+// a chance to sync them.
+type MaxTimeExpandedReplication struct {
+	maxTime     time.Duration
+	gracePeriod time.Duration
+	now         func() time.Time
+}
+
+func (e *MaxTimeExpandedReplication) isMaxTimeInWindow(b ReplicatedBlock, now time.Time) bool {
+	maxTimeDelta := now.Sub(b.GetMaxTime())
+	return maxTimeDelta <= e.maxTime
+}
+
+func (e *MaxTimeExpandedReplication) EligibleForSync(b ReplicatedBlock) bool {
+	return e.isMaxTimeInWindow(b, e.now())
+}
+
+func (e *MaxTimeExpandedReplication) EligibleForQuerying(b QueryableReplicatedBlock) bool {
+	now := e.now()
+	uploadedDelta := now.Sub(b.GetUploadedAt())
+	return uploadedDelta > e.gracePeriod && e.isMaxTimeInWindow(b, now)
```

Comment:
This grace period is for the time the block was uploaded but I think we also need one for the max time? If a block was uploaded hours ago, we still want it to be synced by the store-gateways before the queriers start expecting store-gateways to have it.

Reply:
Fixed via 450dda17db9b00035717da5e2fac05676ab21cbe